### ISSUE TITLE
Different results when decoding with different batch sizes (--max-sentences)

### ISSUE THREAD

**Issue by joyolee** [pytorch/fairseq#3078]
Hi,

I would ask the reason why I get different results when I decode with different batch sizes. (--max-sentences).

I find this issue when I use the espresso toolkit, which is an ASR toolkit and based on the fairseq. I have contacted the owner of Espresso toolkit, he observed the same situation. And we have the same opinion that it might be caused by beam search in batches `sequence_generator.py`. For this reason, I come here to ask if you can make us clear about this question.  

For example, with the same attention-based encoder-decoder model:

1. by setting the batch size (--max-sentences) as 32, the WER is 3.46% in the test set and 5.71% in the validation set.
2. by setting the batch size (--max-sentences) as 1, the WER is 3.42% in the test set and 5.67% in validation set.

The difference is not large, I guess this issue is caused by the beam search in batches. But I'm not sure clearly where the difference comes from. If you know, I'm glad to have your answer.

**Comment by myleott** [comment]
I suspect it's due to the model not zeroing out padding tokens, rather than beam search.

Particularly for encoder-decoder models, in order for the results to be truly independent of batch size, you'd need to have the encoder zero padding symbols after every layer. Otherwise you may get slightly different results with different batch sizes due to having more/fewer padding symbols.

While we do correct for this in [the convolutional model](https://github.com/pytorch/fairseq/blob/7e5e45b483ec5896830231ebbd3ed26472bcff47/fairseq/models/fconv.py#L269-L270), I think the Transformer model does not have this correction, since it slows things down and doesn't usually influence the results much anyway.

Note that the embedding value for the pad symbol is always 0s, but the outputs can become non-zero due to bias terms (e.g., in the FFN).

**Comment by joyolee** [comment]
It makes sense. I'm going to check the transformer code. 

Many thanks for your reply!
